core/execution/engine.go

```diff
@@ -1013,12 +1005,12 @@ func (e *Engine) OnTick(ctx context.Context, t time.Time) {
 	toDelete := []string{}
 	parentStates := e.getParentStates()
 	evts := make([]events.Event, 0, len(e.marketsCpy))
-	rejected := map[int]struct{}{}
+	maxI := len(e.marketsCpy) - 1
 	for i, mkt := range e.marketsCpy {
-		if _, ok := rejected[i]; ok {
-			// successor markets were rejected, because of how golang iterates over slices, this loop
-			// will still iterate over rejected markets, these markets must be skipped
-			continue
+		// when a successor market kicks in, the marketsCpy slice is updated
+		// this loop will still iterate over N elements, even if the slice has shrunk
+		if i >= maxI {
```

Comment:
The more I think the more that this is not right.

Say have markets [0,1,2,3,4,5] where markets 2,3,4 are all successors of market 1.

We enter loop `i==2` and say market 2 comes out of auction so we reject `3,4` and our slice looks like `[0,1,2,5]`. We set `maxI` to be `len(e.marketsCpy) - 1 == 3`.

We start loop `i=3` and `if i >= maxI` is true so we break, and we've just clean forgotten to process market number 5.

Heres it happening in goplayground: https://go.dev/play/p/J2jZgd8YyhX

I think we should stop playing with fire, and just not play around with the slice while we iterate over it. Collect the markets to reject and remove them later.

Reply:
Alright, I've updated the PR. There was still some panic in certain integration tests, apparently the `removeMarket` call in the execution engine did not break out of the loop once the market was found. Not sure if that got accidentally removed, or was a bug that had been there for a while